iroh-net/src/magicsock.rs

```diff
@@ -1225,10 +1213,14 @@ impl MagicSock {
         }
     }
 
+    /// Sends the call-me-maybe DISCO message, queuing if addresses are too stale.
+    ///
+    /// To send the call-me-maybe message we need to know our current direct addresses.  If
+    /// this information is too stale the call-me-maybe is queued while a netcheck run is
+    /// scheduled.  Once this run finishes the call-me-maybe will be sent.
     fn send_or_queue_call_me_maybe(&self, url: &RelayUrl, dst_node: NodeId) {
-        let direct_addrs = self.direct_addrs.read();
-        if direct_addrs.fresh_enough() {
-            let msg = direct_addrs.to_call_me_maybe_message();
+        if self.direct_addrs.fresh_enough() {
```

Comment:
I think this changes the semantics. Before, this method worked on one copy of the direct addresses, now there is a read access per call which can potentially lead to different values. So, tldr I think this is racy. Or at least seen only from the perspective of this function. If there is some magic above preventing multiple accesses to this, I'm not sure

Reply:
Firstly, I don't think this should rely on magic outside of the function so I'm with you with ignoring what happens in the callers.

Looking at the code however I think the only thing which is racy is the debug log of `last_refresh_ago`.  If someone updates the direct addresses after the `fresh_enough` call says it's fresh enough... it still is fresh enough.  So I don't think calling `to_call_me_maybe_messages` separate from calling `fresh_enough` is a problem and am tempted to leave that.

Reply:
I changed the API of `fresh_enough` to return `Result<(), Duration>` to fix this.  Also gets rid of the horrible `last_fresh_ago` method :smile: 

I appreciate this is not the best code.  I my current thinking to do this thoroughly is to split this whole direct address management into an actor and have that actor just keep it's state (splitting off all that state from the magicsock will already be grand!) and refresh this stuff.  Whenever is has new direct addresses it should send them to the magicsock so it can store those for access on the hot path.  that actor itself can basically supervise itself: either it continues to run and provides addresses or it panics and brings down the process - i dont' think there needs to be an in-between (bear in mind "fresh enough" is 27 seconds).

(I also think that the interfaces information should be in an actor that behaves somewhat similar, but that's very tangential to this.)